util/histogram_windowing.cc

```diff
@@ -0,0 +1,193 @@
+}
+
+void HistogramWindowingImpl::Merge(const HistogramWindowingImpl& other) {
+  std::lock_guard<std::mutex> lock(mutex_);
+  stats_.Merge(other.stats_);
+
+  if (stats_.num_buckets_ != other.stats_.num_buckets_ ||
+      micros_per_window_ != other.micros_per_window_) {
+    return;
+  }
+
+  uint64_t cur_window = current_window();
+  uint64_t other_cur_window = other.current_window();
+  // going backwards for alignment
+  for (unsigned int i = 0; 
+                    i < std::min(num_windows_, other.num_windows_); i++) {
+    uint64_t window_index =
+        (cur_window + num_windows_ - i) % num_windows_;
+    uint64_t other_window_index =
+        (other_cur_window + other.num_windows_ - i) % other.num_windows_;
+
+    window_stats_[window_index].Merge(other.window_stats_[other_window_index]);
+  }
+}
+
+std::string HistogramWindowingImpl::ToString() const {
+  return stats_.ToString();
+}
+
+double HistogramWindowingImpl::Median() const {
+  return Percentile(50.0);
+}
+
+double HistogramWindowingImpl::Percentile(double p) const {
+  // Retry 3 times in total
+  for (int retry = 0; retry < 3; retry++) {
+    uint64_t start_num = stats_.num();
+    double result = stats_.Percentile(p);
+    // Detect if swap buckets or Clear() was called during calculation
+    if (stats_.num() >= start_num) {
+      return result;
+    }
+  }
+  return 0.0;
+}
+
+double HistogramWindowingImpl::Average() const {
+  return stats_.Average();
+}
+
+double HistogramWindowingImpl::StandardDeviation() const {
+  return stats_.StandardDeviation();
+}
+
+void HistogramWindowingImpl::Data(HistogramData * const data) const {
+  stats_.Data(data);
+}
+
+void HistogramWindowingImpl::TimerTick() {
+  uint64_t curr_time = env_->NowMicros();
+  if (curr_time - last_swap_time() > micros_per_window_ &&
+      window_stats_[current_window()].num() >= min_num_per_window_) {
+    SwapHistoryBucket();
+  }
+}
+
+void HistogramWindowingImpl::SwapHistoryBucket() {
+  // Threads executing Add() would be competing for this mutex, the first one
+  // who got the metex would take care of the bucket swap, other threads
+  // can skip this.
+  // If mutex is held by Merge() or Clear(), next Add() will take care of the
+  // swap, if needed.
+  if (mutex_.try_lock()) {
+    last_swap_time_.store(env_->NowMicros(), std::memory_order_relaxed);
+
+    uint64_t next_window = (current_window() + 1) % num_windows_;
+
+    // subtract next buckets from totals and swap to next buckets
+    HistogramStat& stats_to_drop = window_stats_[next_window];
+
+    if (!stats_to_drop.Empty()) {
+      for (size_t b = 0; b < stats_.num_buckets_; b++){
+        stats_.buckets_[b].fetch_sub(
+            stats_to_drop.bucket_at(b), std::memory_order_relaxed);
+      }
+
+      if (stats_.min() == stats_to_drop.min()) {
+        uint64_t new_min = bucketMapper.LastValue();
+        for (unsigned int i = 1; i < num_windows_; i++) {
+          uint64_t m = window_stats_[(next_window + i) % num_windows_].min();
```

Comment:
I'm a little bit paranoid but I hope we can avoid using divide here. Instead of num_windows_, can you use num_windows_bits_ here so that we can use bit operation here? Histogram can be invoked in critical paths so I don't want to introduce expensive instruction like divide here unless necessary.